### PR TITLE
feat: Add email template for inviting user to workspace

### LINKED ISSUE #54: Mail template for inviting user to project
## Description

Add a mail template that will be sent to registered users when someone invites them to a project.

## Variables
- project name
- project joining URL
- invited by
- invited on

### PR DESCRIPTION
**Pull request by Allan2000-Git**
## Description

This PR adds an email template for inviting user to project. It provides users with key information regarding the project name
project joining URL, invited by, invited on and invitation role.

Fixes #54 

## Dependencies

No new dependencies introduced.

## Future Improvements

N/A

## Mentions

@rajdip-b 

## Screenshots of relevant screens
workspace-invitation-email

## Developer's checklist

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [x] I have added relevant screenshots in my PR
- [x] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

___

### **PR Type**
enhancement

___

### **Description**
- Added a new `userInvitation` method to the `IMailService` interface, allowing for user invitations via email.
- Implemented the `userInvitation` method in `MailService`, which sends an HTML formatted email containing project invitation details and a call-to-action button.
- Created a mock version of the `userInvitation` method in `MockMailService` for testing, which logs the invitation details.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>interface.service.ts</strong><dd><code>Add user invitation method to mail service interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/mail/services/interface.service.ts

<li>Added <code>userInvitation</code> method to the <code>IMailService</code> interface.<br> <li> Method includes parameters for email, project details, inviter, and <br>role.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/480/files#diff-7bd4bf0fe11fee5f461d6d344313544b6b57ef2c4dedcac37a74538c70f61254">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mail.service.ts</strong><dd><code>Implement user invitation email with HTML template</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/mail/services/mail.service.ts

<li>Implemented <code>userInvitation</code> method in <code>MailService</code>.<br> <li> Constructed HTML email template for project invitation.<br> <li> Included project details and call-to-action button in the email.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/480/files#diff-7eef5c389ba70573686bfbc12ad026cd9a596980cb56c7192d0958d156d5e39a">+184/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mock.service.ts</strong><dd><code>Add mock user invitation method for testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/mail/services/mock.service.ts

<li>Added mock implementation of <code>userInvitation</code> method.<br> <li> Logs invitation details for testing purposes.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/480/files#diff-8fa025332af9f03528e3caaf6c3a1abda133cb0893bb8204ce132366b1995c19">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by rajdip-b** [comment]
Hey man, that's some impressive work. Although again, the scope of this issue is outdated. So it would be Workspace, not project. And gladly, we already have a function for workspace. All you would need is to update the stylings of that mail template.

Additionally, I would like to take a call from @kriptonian1 @darksaiii about the design decisions.

**Comment by kriptonian1** [comment]
@Allan2000-Git I like the work, but can you do the implementation in [React email](https://react.email/), they have better support for all email client and also have accessibility for clients which doesn't support HTML.

**Comment by Allan2000-Git** [comment]
@rajdip-b Will rename project to workspace.
@kriptonian1 I guess all the templates are written in plain HTML & CSS in `mail.service.ts`. Should I change everyone of them or just keep the scope for the current PR?

**Comment by kriptonian1** [comment]
> @rajdip-b Will rename project to workspace. @kriptonian1 I guess all the templates are written in plain HTML & CSS in `mail.service.ts`. Should I change everyone of them or just keep the scope for the current PR?

@rajdip-b should we make a new issue for this

**Comment by rajdip-b** [comment]
A new PR would be really great. I don't think we need a new issue for that, adding hacktoberfest-accepted to the PR should be enough.

**Comment by kriptonian1** [comment]
> A new PR would be really great. I don't think we need a new issue for that, adding hacktoberfest-accepted to the PR should be enough.

Yes, as per as I know, they just count the PR with `hacktoberfest-accepted` tag

**Comment by rajdip-b** [comment]
> Hey man, that's some impressive work. Although again, the scope of this issue is outdated. So it would be Workspace, not project. And gladly, we already have a function for workspace. All you would need is to update the stylings of that mail template.
> 
> Additionally, I would like to take a call from @kriptonian1 @darksaiii about the design decisions.

@Allan2000-Git if you can implement these changes, I will be happy to merge.

**Comment by Allan2000-Git** [comment]
> > Hey man, that's some impressive work. Although again, the scope of this issue is outdated. So it would be Workspace, not project. And gladly, we already have a function for workspace. All you would need is to update the stylings of that mail template.
> > Additionally, I would like to take a call from @kriptonian1 @darksaiii about the design decisions.
> 
> @Allan2000-Git if you can implement these changes, I will be happy to merge.

@rajdip-b Since the template has to be designed using `react-email`, it's bit difficult since I'm not able to preview the code. So would need help there.

**Comment by rajdip-b** [comment]
@kriptonian1 can you please help him out?

**Comment by kriptonian1** [comment]
> > > Hey man, that's some impressive work. Although again, the scope of this issue is outdated. So it would be Workspace, not project. And gladly, we already have a function for workspace. All you would need is to update the stylings of that mail template.
> > > Additionally, I would like to take a call from @kriptonian1 @darksaiii about the design decisions.
> > 
> > @Allan2000-Git if you can implement these changes, I will be happy to merge.
> 
> @rajdip-b Since the template has to be designed using `react-email`, it's bit difficult since I'm not able to preview the code. So would need help there.

React email has preview, I'll recommend you to go through the docs and watch some tutorials on YouTube before diving into it.

Here is a video you can refer to

https://youtu.be/pHXHBHdTzrM?si=BAg-r_EYyJmfhK2W

**Comment by rajdip-b** [comment]
@Allan2000-Git hey buddy, any updates on this?

**Comment by Allan2000-Git** [comment]
@rajdip-b There's already a PR with react email usage. Since most the things are failing, so I didn't go ahead with the same here.

**Comment by Allan2000-Git** [comment]
Template is ready and is working fine

**Comment by Allan2000-Git** [comment]
Can you pull the changes from this [PR](https://github.com/keyshade-xyz/keyshade/pull/476) and check. Would be of great help

File: apps/api/src/mail/services/interface.service.ts [hunk 1]
```diff
@@ -19,6 +19,15 @@ export interface IMailService {
 
   feedbackEmail(email: string, feedback: string): Promise<void>
 
+  userInvitation(
```
Comment (rajdip-b) [inline apps/api/src/mail/services/interface.service.ts:22]: Rename this to `invitedToWorkspace`

**Review by rajdip-b — CHANGES_REQUESTED** [review]

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: apps/api/src/mail/services/interface.service.ts [hunk 2]
```diff
@@ -5,11 +5,12 @@ export interface IMailService {
 
   sendEmailChangedOtp(email: string, otp: string): Promise<void>
 
-  workspaceInvitationMailForUsers(
+  invitedToWorkspace(
     email: string,
-    workspace: string,
-    actionUrl: string,
+    projectName: string,
```
Comment (rajdip-b) [inline apps/api/src/mail/services/interface.service.ts:10]: Not sure why you made this change. Considering the fact that people won't be invited to projects but workspaces.

File: apps/api/src/mail/emails/workspace-invitation.tsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  return (
+    <Html>
+      <Head />
+      <Preview>
+        {forRegisteredUser
+          ? 'Welcome Back! Join Your Workspace'
+          : 'You are Invited to Join the Workspace'}
+      </Preview>
+      <Body style={main}>
+        <Container style={container}>
```
Comment (rajdip-b) [inline apps/api/src/mail/emails/workspace-invitation.tsx:57]: I'm pretty sure we can re-use the frames in here. We can convert this into a reusable component where in we can pass the body, the rest would be pre set.

**Review by rajdip-b — CHANGES_REQUESTED** [review]

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**